### PR TITLE
New port: RapidJSON

### PR DESCRIPTION
**Pull request by Lord-Kamina**
#### Description

RapidJSON is a fast, unicode-friendly, self-contained and header-only library without any dependencies for parsing JSON.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.13.4
Xcode 9.3

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: lang/rapidjson/Portfile [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+PortSystem                      1.0
+PortGroup                       github 1.0
+PortGroup                       cmake 1.1
+
+github.setup                    Tencent rapidjson b32cd9421c5e3cbe183a99b6537ce11441e50656
+version                         20180424
```
Comment (raimue) [inline lang/rapidjson/Portfile]: Why a snapshot from master and not a [release](https://github.com/Tencent/rapidjson/releases)?

File: lang/rapidjson/Portfile [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+depends_build-append            bin:doxygen:doxygen
+
+# Disbled Valgrind because it's apparently failing not due to issues with RapidJSON but rather macOS itself.
+patchfiles                      patch-test-CmakeLists.txt.diff \
+                                patch-test-unittest.diff \
+                                patch-test-unittest-CmakeLists.txt.diff
+
+configure.pre_args-delete       DCMAKE_POLICY_DEFAULT_CMP0025=NEW
+configure.post_args-append      -DRAPIDJSON_HAS_STDSTRING=ON -DRAPIDJSON_HAS_CXX11_RVALUE_REFS=ON -DRAPIDJSON_BUILD_TESTS=OFF
+configure.optflags-delete   -   Os
```
Comment (raimue) [inline lang/rapidjson/Portfile]: Something is off here with the `-`.

File: lang/rapidjson/Portfile [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+PortSystem                      1.0
+PortGroup                       github 1.0
+PortGroup                       cmake 1.1
+
+github.setup                    Tencent rapidjson b32cd9421c5e3cbe183a99b6537ce11441e50656
+version                         20180424
+license                         MIT
+categories                      lang devel
+platforms                       darwin
+maintainers                     @Lord-Kamina
```
Comment (raimue) [inline lang/rapidjson/Portfile]: Should also [include an email address](https://guide.macports.org/#reference.keywords.maintainers), so you can be contacted from other systems by email (for example our buildbot).

File: lang/rapidjson/Portfile [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                                sha256  4133516f8fd578053f478f3b6211ebcfb03056bdd59d6c6d0312c393f467cf37
+homepage                        http://rapidjson.org/
+depends_build-append            bin:doxygen:doxygen
+
+# Disbled Valgrind because it's apparently failing not due to issues with RapidJSON but rather macOS itself.
+patchfiles                      patch-test-CmakeLists.txt.diff \
+                                patch-test-unittest.diff \
+                                patch-test-unittest-CmakeLists.txt.diff
+
+configure.pre_args-delete       DCMAKE_POLICY_DEFAULT_CMP0025=NEW
```
Comment (raimue) [inline lang/rapidjson/Portfile]: This is missing the leading `-`.

File: lang/rapidjson/Portfile [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+cmake.generator                 Unix Makefiles
+
+variant tests description {Build and run unit tests.} {
+    depends_build-append        port:gtest
+    configure.post_args-delete  -DRAPIDJSON_BUILD_TESTS=OFF
+    test {
+        test.run                    yes
+        test.cmd                    make test
+        system -W ${cmake.build_dir} ${test.cmd}
+        }
```
Comment (raimue) [inline lang/rapidjson/Portfile]: This cannot work like this. `test.run` has to be defined outside of the phase. Why are you using system here?

You should only need these two options:
```
test.run yes
test.target test
```

File: lang/rapidjson/Portfile (hunk 1 above)
Comment (Lord-Kamina) [reply lang/rapidjson/Portfile]: Last release is like 2 years old.

File: lang/rapidjson/Portfile (hunk 2 above)
Comment (Lord-Kamina) [reply lang/rapidjson/Portfile]: So it is.

File: lang/rapidjson/Portfile (hunk 3 above)
Comment (Lord-Kamina) [reply lang/rapidjson/Portfile]: I thought it was one or the other, I guess I should correct the other PRs as well.

File: lang/rapidjson/Portfile (hunk 5 above)
Comment (Lord-Kamina) [reply lang/rapidjson/Portfile]: I did it that way the first time I tried and for some reason it did nothing, where as defining it as I did worked fine.

I'm gonna try it again later.

File: lang/rapidjson/Portfile (hunk 1 above)
Comment (raimue) [reply lang/rapidjson/Portfile]: Upstream must have its reasons not to provide a newer release. Ports are supposed to provide stable versions. A prerelease following master could additionally be provided by a port with the`-devel` suffix.

Comment (Lord-Kamina) [reply lang/rapidjson/Portfile]: @raimue whatever the reason, 1.1.0 fails to build most examples and tests, and will not work with c++11.

Comment (raimue) [reply lang/rapidjson/Portfile]: If there is no way to make 1.1.0 work, this port should be named rapidjson-devel to indicate it is not following releases.

Comment (Lord-Kamina) [reply lang/rapidjson/Portfile]: Will do. I spent all day yesterday trying to get 1.1.0 to work and eventually gave up. Obviously, if there's a new release in the future that works, I'd fix the port to use that.

**Comment by wyuenho** [comment]
@Lord-Kamina are you still working on this PR?

**Comment by mojca** [comment]
What's the status of this PR? The above conversation seems to suggest that further commits are on the way? It would be nice to close the oldest PRs. Once I get feedback to feedback to #1711, this will be the oldest one, so time to get it in shape for merging?

**Comment by Lord-Kamina** [comment]
@mojca I don't think I'll keep working on this one after all, I'm looking into Nlohmann instead.

**Comment by kencu** [comment]
Looks good to me. Builds easily, passes all tests. Even if you don't want to maintain it, do you mind if I go ahead and fix up the final few minor issues and commit this?

**Comment by Lord-Kamina** [comment]
@kencu by all means go ahead. Even if I decide to use another JSON parser in the end, somebody will surely find RapidJSON useful.

**Comment by kencu** [comment]
Thanks. LGTM. Hope the buildbot farm likes it too :>

e948e1bb604d7fafa5476b16fa91cb671f591d4e

**Pull request closed without merging**